torchtext/datasets/wikitext2.py

```diff
@@ -25,10 +29,23 @@
 @_add_docstring_header(num_lines=NUM_LINES)
 @_create_dataset_directory(dataset_name=DATASET_NAME)
 @_wrap_split_argument(('train', 'valid', 'test'))
-def WikiText2(root, split):
-    dataset_tar = download_from_url(URL, root=root, hash_value=MD5, hash_type='md5')
-    extracted_files = extract_archive(dataset_tar)
-    path = _find_match(split, extracted_files)
-    logging.info('Creating {} data'.format(split))
-    return _RawTextIterableDataset(DATASET_NAME,
-                                   NUM_LINES[split], _read_text_iterator(path))
+def WikiText2(root: str, split: Union[Tuple[str], str]):
+    if not is_module_available("torchdata"):
+        raise ModuleNotFoundError("Package `torchdata` not found. Please install following instructions at `https://github.com/pytorch/data`")
+    url_dp = IterableWrapper([URL])
+    # cache data on-disk
+    filepath_fn = functools.partial(lambda x: os.path.join(root, os.path.basename(x)))
+    cache_dp = url_dp.on_disk_cache(
+        filepath_fn=filepath_fn,
+        hash_dict={os.path.join(root, os.path.basename(URL)): MD5},
+        hash_type="md5",
+    )
+    cache_dp = HttpReader(cache_dp).end_caching(mode="wb", same_filepath_fn=True)
+    cache_dp = FileOpener(cache_dp, mode="b")
+    # stack Zip extractor on top of load files data pipe
+    extracted_files = cache_dp.read_from_zip()
+    # filter the files as applicable to create dataset for given split (train or test)
+    filter_fn = functools.partial(lambda x: split in Path(x[0]).parts[-1])
+    filter_extracted_files = extracted_files.filter(filter_fn)
+    extract_text_fn = functools.partial(lambda t: t[1].decode())
```

Comment:
IMO this is needed and we should ensure that all our datapipes are free of lambdas. This is because lambdas are not supported by pickle. So when you initialize these datapipes, you will get a warning from PyTorch codebase: https://github.com/pytorch/pytorch/blob/master/torch/utils/data/datapipes/iter/callable.py#L55

Reply:
Understood thanks for letting me know about the pickling issues with lambda functions. We should make a note of this in the GH issue with a link to this PR so future dataset migrations can follow this example.